au4/src/projectscene/view/clipsview/timelineruler.cpp

```diff
@@ -0,0 +1,207 @@
+#include <cassert>
+#include <QtCore/QString>
+#include <QtCore/QVector>
+#include <QtGui/QColor>
+#include <QtGui/QPainter>
+
+#include "timelineruler.h"
+
+namespace {
+constexpr int MINORMINOR_TICK_HEIGHT_RATIO = 8;
+constexpr int MINOR_TICK_HEIGHT_RATIO = 4;
+constexpr int FONT_SIZE = 12;
+constexpr double TICK_ALPHA_MAJOR = 0.6;
+constexpr double TICK_ALPHA_MINOR = 1.0;
+constexpr int LABEL_OFFSET = 3;
+constexpr int LABEL_INTERVAL = 5;
+}
+
+TimelineRuler::TimelineRuler(QQuickItem* parent)
+    : QQuickPaintedItem(parent)
+{
+}
+
+void TimelineRuler::paint(QPainter* painter)
+{
+    assert(painter != nullptr);
+    if (painter == nullptr) {
+        return;
+    }
+
+    const auto w = width();
+    const auto h = height();
+
+    TickInfo tickInfo;
+    QVector<QLineF> ticks;
+    TickLabels labels;
+
+    // determine current time interval and prepare ticks
+    TimeFormat format;
+    format.SetTickSizes(m_zoom, tickInfo.major, tickInfo.minor, tickInfo.minorMinor, tickInfo.digits);
+    prepareTickData(ticks, labels, tickInfo, format, w, h);
+
+    // begin painting
+    painter->save();
+    QPen pen = painter->pen();
+    pen.setWidth(1);
+
+    if (m_textColor.isValid()) {
+        drawLabels(painter, labels, w, h);
+    }
+
+    if (m_tickColor.isValid()) {
+        drawTicks(painter, ticks);
+    }
+
+    painter->restore();
+}
+
+void TimelineRuler::prepareTickData(QVector<QLineF>& ticks, TickLabels& labels, TickInfo& tickInfo, TimeFormat& format, double w, double h)
```

Comment:
there are a lot of objects being passed by non-constant reference, are they all changed by this method?